include/derecho/core/detail/external_group_impl.hpp

```diff
@@ -0,0 +1,436 @@
+        : my_id(getConfUInt32(CONF_DERECHO_LOCAL_ID)),
+          receivers(new std::decay_t<decltype(*receivers)>()) {
+    if(deserialization_context != nullptr) {
+        rdv.push_back(deserialization_context);
+    }
+#ifdef USE_VERBS_API
+    sst::verbs_initialize(member_ips_and_sst_ports_map,
+                          curr_view->members[curr_view->my_rank]);
+#else
+    sst::lf_initialize({}, my_id, std::map<node_id_t, std::pair<ip_addr_t, uint16_t>>{{my_id, 
+    {getConfString(CONF_DERECHO_LOCAL_IP), getConfUInt16(CONF_DERECHO_EXTERNAL_PORT)}}});
+#endif
+
+    get_view(-1);
+
+    uint64_t view_max_rpc_reply_payload_size = 0;
+    uint32_t view_max_rpc_window_size = 0;
+
+    for(subgroup_id_t subgroup_id = 0; subgroup_id < curr_view->subgroup_shard_views.size(); ++subgroup_id) {
+        uint64_t max_payload_size = 0;
+        uint32_t num_shards = curr_view->subgroup_shard_views.at(subgroup_id).size();
+        for(uint32_t shard_num = 0; shard_num < num_shards; ++shard_num) {
+            SubView& shard_view = curr_view->subgroup_shard_views.at(subgroup_id).at(shard_num);
+            const DerechoParams& profile = DerechoParams::from_profile(shard_view.profile);
+            uint64_t payload_size = profile.max_msg_size - sizeof(header);
+            max_payload_size = std::max(payload_size, max_payload_size);
+            view_max_rpc_reply_payload_size = std::max(
+                    profile.max_reply_msg_size - sizeof(header),
+                    view_max_rpc_reply_payload_size);
+            view_max_rpc_window_size = std::max(profile.window_size, view_max_rpc_window_size);
+        }
+        max_payload_sizes[subgroup_id] = max_payload_size;
+    }
+
+    p2p_connections = std::make_unique<sst::P2PConnectionManager>(sst::P2PParams{
+            my_id,
+            getConfUInt32(CONF_DERECHO_P2P_WINDOW_SIZE),
+            view_max_rpc_window_size,
+            getConfUInt64(CONF_DERECHO_MAX_P2P_REPLY_PAYLOAD_SIZE) + sizeof(header),
+            getConfUInt64(CONF_DERECHO_MAX_P2P_REQUEST_PAYLOAD_SIZE) + sizeof(header),
+            view_max_rpc_reply_payload_size + sizeof(header),
+            true,
+            NULL});
+
+    rpc_thread = std::thread(&ExternalGroup<ReplicatedTypes...>::p2p_receive_loop, this);
+}  // namespace derecho
+
+template <typename... ReplicatedTypes>
+ExternalGroup<ReplicatedTypes...>::~ExternalGroup() {
+    thread_shutdown = true;
+    if(rpc_thread.joinable()) {
+        rpc_thread.join();
+    }
+}
+
+template <typename... ReplicatedTypes>
+bool ExternalGroup<ReplicatedTypes...>::get_view(const node_id_t nid) {
+    try {
+        tcp::socket sock = curr_view ? 
+        tcp::socket(std::get<0>(curr_view->member_ips_and_ports[curr_view->rank_of(nid)]),
+                    std::get<PORT_TYPE::GMS>(curr_view->member_ips_and_ports[curr_view->rank_of(nid)]),
+                    false) :
+        tcp::socket(getConfString(CONF_DERECHO_LEADER_IP), getConfUInt16(CONF_DERECHO_LEADER_GMS_PORT));
```

Comment:
I understand the goal of this behavior (if there is no current view, default to using the configured Derecho leader, since there's no way to map node IDs to IPs without a view), but it makes the interface to `get_view` a little confusing: The `nid` parameter will sometimes be ignored for reasons not clear to the caller.